lms/product/d2l/_plugin/grouping.py

```diff
@@ -0,0 +1,80 @@
+from enum import Enum
+
+from lms.models import Grouping
+from lms.product.plugin.grouping_service import GroupError, GroupingServicePlugin
+from lms.services import D2LAPIClient
+from lms.services.exceptions import ExternalRequestError
+
+
+class ErrorCodes(str, Enum):
+    """Error codes that the FE is going to check for."""
+
+    GROUP_SET_NOT_FOUND = "d2l_group_set_not_found"
+    GROUP_SET_EMPTY = "d2l_group_set_empty"
+    STUDENT_NOT_IN_GROUP = "d2l_student_not_in_group"
+
+
+class D2LGroupingPlugin(GroupingServicePlugin):
+    """A plugin which implements D2L specific grouping functions."""
+
+    group_type = Grouping.Type.D2L_GROUP
+    sections_type = None  # We don't support sections in D2L
+
+    def __init__(self, d2l_api, api_user_id):
+        self._d2l_api = d2l_api
+        self._api_user_id = api_user_id
+
+    def get_groups_for_learner(self, _svc, course, group_set_id):
+        if learner_groups := self._d2l_api.group_set_groups(
+            course.lms_id, group_set_id, self._api_user_id
+        ):
+            return learner_groups
+
+        raise GroupError(ErrorCodes.STUDENT_NOT_IN_GROUP, group_set=group_set_id)
+
+    def get_groups_for_grading(
+        self, svc, course, group_set_id, grading_student_id=None
+    ):
+        return svc.get_course_groupings_for_user(
+            course,
+            grading_student_id,
+            type_=self.group_type,
+            group_set_id=int(group_set_id),
+        )
+
+    def get_groups_for_instructor(self, _svc, course, group_set_id):
+        try:
+            groups = self._d2l_api.group_set_groups(course.lms_id, group_set_id)
+        except ExternalRequestError as exc:
+            if exc.status_code == 404:
+                raise GroupError(
+                    ErrorCodes.GROUP_SET_NOT_FOUND, group_set=group_set_id
+                ) from exc
+
+            raise
+
+        if not groups:
+            raise GroupError(ErrorCodes.GROUP_SET_EMPTY, group_set=group_set_id)
+
+        return groups
+
+    @staticmethod
+    def _get_api_user_id(user_id: str):
```

Comment:
Here's the biggest LMS specific wart in all of this.

I reckon the right solution is to abstract this inside LTIUser as `api_lti_user` or similar and handle it as a plugin or similar for Canvas at the same time.

Added a note to the board about that.

On this PR this is relatively hidden for the rest of the code. I have it here but you could argue it belongs closer to the API client.

In the canvas case one ID can't be derived from the other so we can't have a translation function in that case.